network/peer.go

```diff
@@ -65,20 +73,176 @@ type peer struct {
 	lastSent, lastReceived int64
 
 	tickerCloser chan struct{}
-
-	// ticker processes
-	tickerOnce sync.Once
 }
 
 // assume the stateLock is held
-func (p *peer) Start() {
+func (p *peer) Start() error {
+	if err := p.conn.SetReadDeadline(p.net.clock.Time().Add(p.net.readPeerVersionTimeout)); err != nil {
+		p.net.log.Verbo("error on setting the connection read timeout %s", err)
+		return err
+	}
+
+	// send the version and get a msg from receiver..
+	msg, err := p.verionAck()
+	if err != nil {
+		return err
+	}
+
+	switch msg.Op() {
+	case PeerList:
+		// handle old self check logic
+		if p.id.Equals(p.net.id) {
+			return errPeerIsMyself
+		}
+
+		// if the first message is not version.  fall back to normal processing logic.
+		// only acceptable option at this point would be a PeerList request
+		// lets take care of the peer request
+		go p.handle(msg)
+
+		go p.ReadMessages()
+
+		// we need to ask for a version.
+		go p.Version()
+		go p.WriteMessages()
+
+		go p.requestFinishHandshake()
+		go p.sendPings()
+
+		return nil
+	case Version:
+	// fallthrough
+	default:
+		// handle old self check logic
+		if p.id.Equals(p.net.id) {
+			return errPeerIsMyself
+		}
+
+		// We didn't get a Version msg.  This is unexpected..
+		return errVersionExpected
+	}
+
+	// parse and check if version is correct.
+	peerVersionStr := msg.Get(VersionStr).(string)
+	var peerVersion version.Version
+	peerVersion, err = p.net.parser.Parse(peerVersionStr)
```

Comment:
nit: To save a line you can do `peerVersion, err := p.net.parser.Parse(peerVersionStr)`